pkg/integrations/v4/dm/id_provider.go

```diff
@@ -4,102 +4,151 @@
+func newUnregisteredEntity(entity protocol.Entity, reason reason, err error) unregisteredEntity {
+	return unregisteredEntity{
 		Entity: entity,
 		Reason: reason,
 		Err:    err,
 	}
 }
 
 type idProviderInterface interface {
-	Entities(agentIdn entity.Identity, entities []protocol.Entity) (registeredEntities RegisteredEntitiesNameToID, unregisteredEntities UnregisteredEntities)
+	ResolveEntities(entities []protocol.Entity) (registeredEntitiesNameToID, unregisteredEntityListWithWait)
 }
 
-type idProvider struct {
+type cachedIdProvider struct {
 	client               identityapi.RegisterClient
-	cache                RegisteredEntitiesNameToID
-	unregisteredEntities UnregisteredEntitiesNamed
+	agentIdentity        func() entity.Identity
+	cache                registeredEntitiesNameToID
+	unregisteredEntities unregisteredEntitiesNamed
+	cacheMutex           *sync.Mutex
+	toRegisterQueue      chan entityListToRegisterWithWait
 }
 
-func NewIDProvider(client identityapi.RegisterClient) *idProvider {
-	cache := make(RegisteredEntitiesNameToID)
-	unregisteredEntities := make(UnregisteredEntitiesNamed)
-	return &idProvider{
+func NewCachedIDProvider(client identityapi.RegisterClient, agentIdentity func() entity.Identity, closeContext context.Context) *cachedIdProvider {
+	cache := make(registeredEntitiesNameToID)
+	unregisteredEntities := make(unregisteredEntitiesNamed)
+	cacheMutex := &sync.Mutex{}
+	provider := cachedIdProvider{
 		client:               client,
+		agentIdentity:        agentIdentity,
 		cache:                cache,
 		unregisteredEntities: unregisteredEntities,
+		cacheMutex:           cacheMutex,
+		toRegisterQueue:      make(chan entityListToRegisterWithWait, 10), // TODO adjust buffer
 	}
+
+	go provider.registerEntitiesWorker(closeContext)
+
+	return &provider
 }
 
-func (p *idProvider) Entities(agentIdn entity.Identity, entities []protocol.Entity) (registeredEntities RegisteredEntitiesNameToID, unregisteredEntities UnregisteredEntities) {
-	unregisteredEntities = make(UnregisteredEntities, 0)
-	registeredEntities = make(RegisteredEntitiesNameToID, 0)
+func (p *cachedIdProvider) ResolveEntities(entities []protocol.Entity) (registeredEntitiesNameToID, unregisteredEntityListWithWait) {
+	unregisteredEntities := make(unregisteredEntityList, 0)
+	registeredEntities := make(registeredEntitiesNameToID, 0)
+
 	entitiesToRegister := make([]protocol.Entity, 0)
 
+	// add error cache checking
+	p.cacheMutex.Lock()
 	for _, e := range entities {
 		if foundID, ok := p.cache[e.Name]; ok {
 			registeredEntities[e.Name] = foundID
+		} else if foundUnregisteredEntities, ok := p.unregisteredEntities[e.Name]; ok {
+			unregisteredEntities = append(unregisteredEntities, foundUnregisteredEntities)
 		} else {
+			unregisteredEntities = append(unregisteredEntities, newUnregisteredEntity(e, reasonNotInCache, nil))
 			entitiesToRegister = append(entitiesToRegister, e)
 		}
 	}
+	p.cacheMutex.Unlock()
+
+	wg := &sync.WaitGroup{}
```

Comment:
I think we can simplify this component and not use the WaitGroup

Reply:
can we have a chat on this?

Reply:
Can also be done with a Channel as well. All subjective though if you choose a WaitGroup or Channel to feed back the information to where we are waiting for the single result.